pyccel/ast/headers.py

```diff
@@ -667,11 +710,70 @@ def apply(self, args, results=None):
             for i, arg in d_arguments.items():
                 if isinstance(arg, Macro):
                     d_arguments[i] = construct_macro(arg.name,
-                                      d_arguments[arg.argument])
+                                                     d_arguments[arg.argument])
                     if isinstance(arg, MacroShape):
                         d_arguments[i]._index = arg.index
+        return d_arguments
+
+
+    def get_results_shapes(self, args):
+        """replace elements of the shape with appropriate values
+
+        Parameters
+        ----------
+        args: iterable
+            The arguments provided at function call
 
+        Results
+        -------
+        results_shapes: iterable
+            List of shapes after replacing variables indicated
+            in the macro if they exist, with the appropriate variables
+            from args.
+        """
+
+        d_arguments = self.link_args(args)
+        argument_keys = d_arguments.keys()
+        results_shapes = []
+        for j, shape in enumerate(self.results_sh):
+            newargs = []
+
+            for i, arg in enumerate(shape):
+                if str(arg) == ':':
+                    try:
+                        new = MacroShape(d_arguments[self.results[j]], i)
+                    except KeyError:
+                        msg = "Shape needs to be provided explicitly as it cannot be deduced"
+                        errors.report(msg, symbol=self.results[j],
+                                      severity='error')
+
+                elif isinstance(arg, PyccelSymbol):
+                    if arg in argument_keys:
+                        new = d_arguments[arg]
+                    else:
+                        new = arg
+                elif isinstance(arg, Macro):
+                    if arg.argument in argument_keys:
+                        new = d_arguments[arg.argument]
+                    else:
+                        raise ValueError('Unknown variable name')
+
+                    if isinstance(arg, MacroShape):
+                        new = construct_macro(arg.name, new, arg.index)
```

Comment:
Is this not clearer if you call `MacroShape` directly ;)

Reply:
I was patiently waiting to steal from your branch h
Fixed here 9e3988d

Reply:
Haha, mine got stuck waiting for someone to take the time to look at it. But as this is new code I think it will cause a collision anyway